pandas/plotting/_matplotlib/tools.py

```diff
@@ -32,19 +32,19 @@ def format_date_labels(ax: "Axes", rot):
 
 def table(ax, data: FrameOrSeries, rowLabels=None, colLabels=None, **kwargs) -> "Table":
     if isinstance(data, ABCSeries):
-        data = data.to_frame()
+        frame = data.to_frame()
     elif isinstance(data, ABCDataFrame):
-        pass
+        frame = data
     else:
         raise ValueError("Input data must be DataFrame or Series")
 
     if rowLabels is None:
-        rowLabels = data.index
+        rowLabels = frame.index
 
     if colLabels is None:
-        colLabels = data.columns
+        colLabels = frame.columns
 
-    cellText = data.values
+    cellText = frame.values
```

Comment:
the changes in this module are not required. The mypy issue is trying to assign the result of data._to_frame back to data which is FrameOrSeries, a typevar.

just need to change FrameOrSeries to FrameOrSeriesUnion in the function signature.

I think the defaults in pandas._typing should be changed to Unions. see also https://github.com/pandas-dev/pandas/pull/36092#issuecomment-724880269 about ArrayLike and AnyArrayLike